### PR TITLE
Add TBB support

### PR DESCRIPTION
**Pull request by costashatz**
This pull-request adds [TBB](https://www.threadingbuildingblocks.org/) support for parallelization. Overall TBB in my experience (with the libcmaes library, but also in general) has super low overhead and provides very nice parallelization. In more detail:

- CMake and the Makefiles should be working (at least in my computer).
- When using CMake one can either use OpenMP or TBB (the user can choose, but not both). TBB is by default ON and OpenMP by default OFF (we can change this). When using Makefiles (which I am not an expert -- on the contrary I am really bad), OpenMP is enabled by default but if TBB is found it will be the one used.

There are the following issues:

- When uncertainty handling is enabled, right now I am not evaluating the candidates in parallel. In my experiments, it doesn't have any big impact in computation speed (only if the evaluation is extremely costly).
- Because enabling the parallelization is a user parameter (`_mt_eval`), I had to duplicate a small part of the code in `ESOStrategy` (more specifically look at the new functions: `candidate_evaluation` and `parallel_candidate_evaluation`).
- In the same note as above, I should have replicated some code in `genopheno.h` (to test for the case `if (candidates.cols() >= 100)`), but in my experiments TBB has very low overhead and should not be a huge problem (also, I didn't want to copy-paste too much code).
- Lastly, currently I have no unit test because it's not easy to test, but I will try to run the BBOB 2013 benchmarks that are available so that I can report some scores.

Let me know what you think...

### CODE REVIEW

**Comment by costashatz** [comment]
@beniz I think I will also improve a bit your `.travis.yml` in this PR (to add some tests with and without TBB). Is there anything specific you would like me not to change? For example, I see you are using very specific compiler version and eigen version.. Do you have any specific reason for that? Now in travis, you can select specific distributions/versions of the OS..

**Comment by beniz** [comment]
Hi, thanks it looks great, though I need to look at what you have pointed to in more depth. One sure point though, I'd recommend you leave OpenMP as default so that nothing breaks or changes other users' setups. I agree that it's fine to refactor the functions in order to work around the pragmas and offer the TBB alternative.

Regarding travis, it's as you like, it could be easier to schedule it as another, complementary PR. As far as I remember Travis comes with a limited number of platforms, and thus I may have pinned some of the packages versions to avoid some issues.

**Comment by beniz** [comment]
Btw, the Travis error looks like it's due to Travis itself:

```
Failed to fetch http://ppa.launchpad.net/chris-lea/redis-server/ubuntu/dists/precise/InRelease  Could not connect to ppa.launchpad.net:80 (91.189.95.83), connection timed out
```

**Comment by costashatz** [comment]
> One sure point though, I'd recommend you leave OpenMP as default so that nothing breaks or changes other users' setups.

Sounds good. I will change to OpenMP as default in CMake. I will also put an option in Makefile to enable or not tbb (even if it is available in the system).

> Regarding travis, it's as you like, it could be easier to schedule it as another, complementary PR.

OK. I will make a new..

> Btw, the Travis error looks like it's due to Travis itself

Yes I saw that, but still it'd be nice to have some tests when using TBB and when not..

File: CMakeLists.txt [hunk 1]
```diff
@@ -70,6 +75,17 @@ if (USE_OPENMP)
...
 
+set(LIBCMAES_DEPS "")
+
+if (USE_TBB)
+  find_package (TBB)
+  if (TBB_FOUND)
+    add_definitions (-DUSE_TBB)
+    include_directories (${TBB_INCLUDE_DIRS})
+    # should use ${TBB_LIBRARIES} but it is not working properly
+    set (LIBCMAES_DEPS "tbb")
```
Comment (jschueller) [inline CMakeLists.txt]: why cannot you use TBB_LIBRARIES ?

File: CMakeLists.txt [hunk 2]
```diff
@@ -70,6 +75,17 @@ if (USE_OPENMP)
   endif ()
 endif ()
 
+set(LIBCMAES_DEPS "")
+
+if (USE_TBB)
+  find_package (TBB)
+  if (TBB_FOUND)
+    add_definitions (-DUSE_TBB)
```
Comment (jschueller) [inline CMakeLists.txt]: usually we define HAVE_XXX

Comment (jschueller) [inline CMakeLists.txt]: maybe you could add a per-target include

File: CMakeLists.txt [hunk 3]
```diff
@@ -70,6 +75,17 @@ if (USE_OPENMP)
   endif ()
 endif ()
 
+set(LIBCMAES_DEPS "")
+
+if (USE_TBB)
+  find_package (TBB)
+  if (TBB_FOUND)
+    add_definitions (-DUSE_TBB)
+    include_directories (${TBB_INCLUDE_DIRS})
```
Comment (jschueller) [inline CMakeLists.txt]: maybe you could add a per-target include

File: CMakeLists.txt (hunk 1 above)
Comment (costashatz) [reply CMakeLists.txt]: > why cannot you use TBB_LIBRARIES ?

To be honest, I do not remember. But the only TBB library is "tbb" so there should not be any issue.

File: CMakeLists.txt (hunk 2 above)
Comment (costashatz) [reply CMakeLists.txt]: > usually we define HAVE_XXX

I will make it HAVE_TBB..

Comment (costashatz) [reply CMakeLists.txt]: > maybe you could add a per-target compile definition

You mean to produce 2 library targets:

- one without TBB
- and one with TBB (if TBB is available)

Otherwise, I do not really see what would be the difference (other than just using a different CMake commad); the library has quite a few targets and `add_definitions` is easier to use in this case (and does not really pollute anything, does it?).

Comment (jschueller) [reply CMakeLists.txt]: no, I mean define the symbol just for this target, ie not define it for tests, python, that's cleaner

Comment (costashatz) [reply CMakeLists.txt]: > no, I mean define the symbol just for this target, ie not define it for tests, python, that's cleaner

But we do want it for the tests (I am not sure for python); I have it in my todo list to make a new PR with additional setups in travis to check for the usage of TBB.

**Comment by beniz** [comment]
@jschueller @costashatz hi guys, thanks for the work on this, I'll keep an eye on it, let me know if you need me to test anything or feedback, and then when you believe it's ready I'll look it up for a merge.

File: CMakeLists.txt (hunk 2 above)
Comment (costashatz) [reply CMakeLists.txt]: @jschueller you mean for the `add_definitions`, right?

Comment (costashatz) [reply CMakeLists.txt]: @jschueller you mean for the add_definitions, right? Or for the CMakelists if?

**Comment by costashatz** [comment]
@beniz This PR should be ready to merge.. Please have a look when you can..

**Comment by beniz** [comment]
@costashatz can you make sure this PR is rebased to the last master ? Thanks.

**Comment by costashatz** [comment]
> @costashatz can you make sure this PR is rebased to the last master ? Thanks.

Done!

File: CMakeLists.txt (hunk 1 above)
Comment (jschueller) [reply CMakeLists.txt]: It can be tbb_static too, see windows binaries.

Comment (costashatz) [reply CMakeLists.txt]: > It can be tbb_static too, see windows binaries.

I do not have a windows machine to test. You're free to test it and make the appropriate changes. Thanks!

Comment (jschueller) [reply CMakeLists.txt]: It should be TBB_LIBRARIES.

File: python/CMakeLists.txt [hunk 4]
```diff
@@ -6,9 +6,9 @@ include_directories (${NUMPY_INCLUDE_DIRS})
 include_directories (${Boost_INCLUDE_DIRS})
 
 python_add_module (lcmaes lcmaes.cc)
-target_link_libraries(lcmaes cmaes ${Boost_LIBRARIES})
+target_link_libraries(lcmaes cmaes ${Boost_LIBRARIES} ${LIBCMAES_DEPS})
 if (LINK_PYTHON)
-  target_link_libraries(lcmaes ${PYTHON_LIBRARIES})
+  target_link_libraries(lcmaes ${PYTHON_LIBRARIES} ${LIBCMAES_DEPS})
 endif ()
```
Comment (jschueller) [inline python/CMakeLists.txt]: It should not be necessary to link to tbb here.

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,168 +1,186 @@
-cmake_minimum_required (VERSION 2.8)
```
Comment (jschueller) [inline CMakeLists.txt]: Please rebase instead of merging, as we see changes from master here.

**Comment by jschueller** [comment]
At least this should be tested, you should enable it in travis & appveyor cis.

File: CMakeLists.txt [hunk 6]
```diff
@@ -1,168 +1,186 @@
...
+find_package (Eigen3 NO_MODULE QUIET)
+if (EIGEN3_FOUND)
+  include (${EIGEN3_USE_FILE})
+  message(STATUS "Eigen3 found: ${EIGEN3_INCLUDE_DIRS}")
+else ()
+  find_package (Eigen3 REQUIRED)
+  include_directories (${EIGEN3_INCLUDE_DIRS})
+endif()
+
+if (USE_OPENMP AND USE_TBB)
```
Comment (jschueller) [inline CMakeLists.txt]: USE_TBB could be on, but not checked if omp is already available, or hte other way around.

**Comment by costashatz** [comment]
> At least this should be tested, you should enable it in travis & appveyor cis.

I am quoting @beniz's reply:

> Regarding travis, it's as you like, it could be easier to schedule it as another, complementary PR.

So, I will handle this in a different PR.

File: python/CMakeLists.txt (hunk 4 above)
Comment (costashatz) [reply python/CMakeLists.txt]: > It should not be necessary to link to tbb here.

It is. Because `lcmaes.cc` includes `cmaes.h` and we have the definition `USE_TBB` enabled. Nevertheless, I agree with you that using TBB inside python (or even OpenMP) is not a good idea. We could write it in the readme that if you want to use the python bindings, disable TBB and OpenMP.

File: CMakeLists.txt (hunk 6 above)
Comment (costashatz) [reply CMakeLists.txt]: > USE_TBB could be on, but not checked if omp is already available, or hte other way around.

This should be fixed now..

File: CMakeLists.txt (hunk 1 above)
Comment (costashatz) [reply CMakeLists.txt]: This should be fixed now..

File: CMakeLists.txt (hunk 2 above)
Comment (jschueller) [reply CMakeLists.txt]: In the CMakeLists, then you'll have to change it in your code.

Comment (costashatz) [reply CMakeLists.txt]: > In the CMakeLists, then you'll have to change it in the cxx files.

I do not get what you mean. Do you mean:

(a)  `add_definitions(-DUSE_TBB)` to be changed to `add_definitions(-DHAVE_TBB)`, or
(b) `if (USE_TBB)` to be changed to `if (HAVE_TBB)`?

File: python/CMakeLists.txt (hunk 4 above)
Comment (costashatz) [reply python/CMakeLists.txt]: > It should not be necessary to link to tbb here.

You're right.. I will remove it..

Comment (jschueller) [reply python/CMakeLists.txt]: Then we should make the use of tbb private: dont expose it in public headers but move the definitions that use tbb in cpp files if possible.
Anyways you're linking twice here in the if (LINK_PYTHON).

**Comment by jschueller** [comment]
We should make the use of tbb private: dont expose it in public headers but move the definitions that use tbb in cpp files.

File: CMakeLists.txt (hunk 5 above)
Comment (costashatz) [reply CMakeLists.txt]: > Please rebase instead of merging, as we see changes from master here

I rebased, but still the changes are not appearing well.

**Comment by costashatz** [comment]
@beniz @jschueller I think I addressed every comment that you made. Thanks for the help and the effort!

In short:

- We can enable TBB both in CMake and Makefiles (with `-DUSE_TBB=ON -DUSE_OPENMP=OFF` in CMake and with `--enable-tbb` in Makefiles)
- I added a test executable where the multi-threading is enabled (`par-test.cc`).
- I will make a new PR for creating some more tests in Travis and AppVeyor to test for TBB and OpenMP parallelization.
- Overall, I tried to insert the minimum code and changes possible. In addition, if TBB is disabled, there should be no change in the behavior of the library.

One thing that remains (irrelevant of whether using TBB or OpenMP), when multi-threaded is enabled in Python the program crashes..

Let me know if you have any other remarks...

File: CMakeLists.txt [hunk 7]
```diff
@@ -59,7 +59,7 @@ check_include_file (sys/types.h HAVE_SYS_TYPES_H)
 check_include_file (sys/stat.h HAVE_SYS_STAT_H)
 
 # eigen
-find_package (Eigen3 NO_MODULE)
+find_package (Eigen3 NO_MODULE QUIET)
```
Comment (jschueller) [inline CMakeLists.txt]: this has nothing to do with tbb support and should be in a separate PR, please describe what you're attempting to fix

Comment (costashatz) [reply CMakeLists.txt]: > this has nothing to do with tbb support and should be in a separate PR, please describe what you're attempting to fix

In the machines that I tried (Ubuntu 15.10 and 16.04), if I do not put `QUIET` I get fatal errors for the cmake configure process. On the contrary, putting the `QUIET` does not influence the behavior of the cmake procedure, because:

- If Eigen3 can be found with the `NO_MODULE` option, then `EIGEN3_FOUND` is set and the check in the following [line](https://github.com/resibots/libcmaes/blob/2ddb3954f735db0340b256ee5a33a060bd523f98/CMakeLists.txt#L63) is true and Eigen3 is properly included.
- If Eigen3 cannot be found with the `NO_MODULE` option or if cmake outputs a *quiet* error, then `EIGEN3_FOUND` is **NOT** set and as such the [else statement](https://github.com/resibots/libcmaes/blob/2ddb3954f735db0340b256ee5a33a060bd523f98/CMakeLists.txt#L66-L69) is executed and the Eigen3 package is searched (note that in the else statement, the Eigen package is `REQUIRED`).

Comment (jschueller) [reply CMakeLists.txt]: Regardless, this is another PR.

**Comment by costashatz** [comment]
> Regardless, this is another PR.

I will do it in another PR then! **Let's merge...**

**Comment by jschueller** [comment]
- You should not use tbb in public headers of libcmaes, it should be only used in code definitions (the functions must be moved to cpp files) so that there's no need for the third party lib to define HAVE_TBB and link to tbb if libcmaes was tbb-enabled

- Please dont include merge commits but rebase your commits on master, as it's difficult to isolate your changes in the global diff.

**Comment by costashatz** [comment]
> You should not use tbb in public headers of libcmaes, it should be only used in code definitions (the functions must be moved to cpp files) so that there's no need for the third party lib to define HAVE_TBB and link to tbb if libcmaes was tbb-enabled

This is not possible as it is now. There is parallelization in `genopheno.h`, which is a public header and cannot be split to a source file because of the templates. I could remove the parallelization there (which should be ok, since generally the overhead in the bound/scale strategies should be small). What do you think @beniz @jschueller? I am ok with removing the parallelization there and making libcmaes with TBB more transparent.

> Please dont include merge commits but rebase your commits on master, as it's difficult to isolate your changes in the global diff.

I rebased.. I do not know why it continues to show it like this. You can always do a [git diff](https://git-scm.com/docs/git-diff).

File: tests/par-test.cc [hunk 8]
```diff
@@ -37,6 +43,9 @@ FitFunc fsphere = [](const double *x, const int N)
 
 int main(int argc, char *argv[])
 {
+#ifdef HAVE_TBB
+  static tbb::task_scheduler_init init;
+#endif
```
Comment (jschueller) [inline tests/par-test.cc]: Shouldnt that go somewhere in cmaes lib rather than in the test ?

Comment (costashatz) [reply tests/par-test.cc]: > Shouldnt that go somewhere in cmaes lib rather than in the test ? Here third party libs would still need to link to tbb.

Not really. Quoting Intel's [documentation](https://software.intel.com/en-us/node/506296):

``Using task_scheduler_init is optional in Intel® Threading Building Blocks (Intel® TBB) 2.2. By default, Intel TBB 2.2 automatically creates a task scheduler the first time that a thread uses task scheduling services and destroys it when the last such thread exits.``

I just put it for completeness.

Comment (jschueller) [reply tests/par-test.cc]: Does it really work without ?

Comment (costashatz) [reply tests/par-test.cc]: > Does it really work without ?

Yes.. You can try it..

Comment (costashatz) [reply tests/par-test.cc]: Yes.. I tried..

Comment (jschueller) [reply tests/par-test.cc]: Ok

File: CMakeLists.txt (hunk 5 above)
Comment (jschueller) [reply CMakeLists.txt]: Its the "TBB almost ok" that is incorrect: it includes commits from master.

Comment (jschueller) [reply CMakeLists.txt]: Maybe "dos2unix CMakeLists.txt" will revert crlfs ?

Comment (costashatz) [reply CMakeLists.txt]: Should be better now..

File: CMakeLists.txt [hunk 9]
```diff
@@ -83,6 +98,8 @@ if (CMAKE_VERSION VERSION_LESS 3.1 OR NOT USE_COMPILE_FEATURES)
   endif ()
 endif ()
 
+set (CMAKE_CXX_FLAGS "-O3 -g -march=native ${CMAKE_CXX_FLAGS}")
```
Comment (jschueller) [inline CMakeLists.txt]: This is unrelated to tbb.

File: CMakeLists.txt [hunk 10]
```diff
@@ -75,6 +76,20 @@ if (USE_OPENMP)
...
 endif ()
 
+set (LIBCMAES_DEPS "")
+set (TBB_DEFS "")
+
+if (USE_TBB)
+  find_package (TBB)
+  if (TBB_FOUND)
+    set (TBB_DEFS "-DHAVE_TBB")
+    set (LIBCMAES_DEPS "${TBB_LIBRARIES}")
```
Comment (jschueller) [inline CMakeLists.txt]: It's useless to define LIBCMAES_DEPS as it's conditionally linked anyways, so you can use TBB_LIBRARIES directly.

File: src/genopheno.h [hunk 11]
```diff
@@ -100,13 +106,20 @@ namespace libcmaes
       if (!_id)
 	{
 	  dMat ncandidates = dMat(candidates.rows(),candidates.cols());
+#ifdef HAVE_TBB
+    tbb::parallel_for(size_t(0), size_t(candidates.cols()), size_t(1), [&](size_t i) {
+#else
```
Comment (jschueller) [inline src/genopheno.h]: These definitions should go in a cpp file.

File: src/surrogates/rankingsvm.hpp [hunk 12]
```diff
@@ -36,6 +36,12 @@
 #include <random>
 #include <iostream>
 
+#ifdef HAVE_TBB
+// Quick hack for definition of 'I' in <complex.h>
+#undef I
+#include <tbb/parallel_for.h>
+#endif
+
```
Comment (jschueller) [inline src/surrogates/rankingsvm.hpp]: These too.

File: tests/par-test.cc [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+  std::vector<double> x0(dim,10.0);
+  double sigma = 0.1;
+  //int lambda = 100; // offsprings at each generation.
+  CMAParameters<> cmaparams(x0,sigma);
+  cmaparams.set_mt_feval(true);
+  //cmaparams._algo = BIPOP_CMAES;
+  CMASolutions cmasols = cmaes<>(fsphere,cmaparams);
+  std::cout << "best solution: " << cmasols << std::endl;
+  std::cout << "optimization took " << cmasols.elapsed_time() / 1000.0 << " seconds\n";
+}
```
Comment (jschueller) [inline tests/par-test.cc]: return 0 ?

File: CMakeLists.txt (hunk 9 above)
Comment (costashatz) [reply CMakeLists.txt]: > This is unrelated to tbb.

It is not, but I had branched from the master that this existed. If you remember we had a discussion about this (#161). Note that this way is the proper way of adding the flags (`-march=native`). We could merge them both and not have different PRs.

File: CMakeLists.txt (hunk 10 above)
Comment (costashatz) [reply CMakeLists.txt]: It's useless to fight that much for things that are harmless. Will change them nevertheless..

File: src/genopheno.h (hunk 11 above)
Comment (costashatz) [reply src/genopheno.h]: I am expecting @beniz opinion on this, because as I already said: These definitions **CANNOT** go to a cpp file; the genopheno class is templated! So either we remove TBB completely from this file (and have no TBB parallelization there) or we leave it and put an explicit remark for users to defive HAVE_TBB if they want TBB. I am OK with both..

File: tests/par-test.cc (hunk 13 above)
Comment (costashatz) [reply tests/par-test.cc]: > return 0 ?

All other tests have the same structure. Not returning anything; I kept it for this one too.

File: CMakeLists.txt (hunk 9 above)
Comment (jschueller) [reply CMakeLists.txt]: I dont agree, this unrelated and should be part of another PR.
This is not proper at all, someone would want to set completely different flags depending on the use case: I think these should not be touched.
Plus their're not accepted by all compilers.

Comment (costashatz) [reply CMakeLists.txt]: > I dont agree, this unrelated and should be part of another PR.

Will make another PR.

> This is not proper at all, someone would want to set completely different flags depending on the use case

Let's make the library header only then. It's the only way of letting the user set the different flags that they want.

> Plus their're not accepted by all compilers.

`-march=native` is accepted by GCC and Clang. I can make a version for ICC (which has `-mtune=native`). In the new PR that I will make, you could test also for Visual Studio (although AppVeyor does not complain at all).

Comment (jschueller) [reply CMakeLists.txt]: No need for header only: leave cmake defaults so that the user can set CMAKE_CXX_FLAGS to whatever he wants. A lot of compiler exist that are different from gcc/clang. Old versions of msvc accept unknown flags but emit a warning. I think more recent versions reject them.

Comment (costashatz) [reply CMakeLists.txt]: > No need for header only: leave cmake defaults so that the user can set CMAKE_CXX_FLAGS to whatever he wants. A lot of compiler exist that are different from gcc/clang. Old versions of msvc accept unknown flags but emit a warning. I think more recent versions reject them.

I removed everything related to optimized flags from this PR. Will leave it for another and we can continue this discussion there..

**Comment by costashatz** [comment]
@beniz @jschueller I thought about it and I am removing TBB from genopheno (the speedup gain would be very small) and from rankingsvm (this is not heavily used and the speedup gain would again not be very big). Let's merge this PR..

**Comment by jschueller** [comment]
I think it should be tested first. Could you at least add libtbb-dev to the debian packages installed in .travis.yml ?

File: tests/par-test.cc (hunk 13 above)
Comment (jschueller) [reply tests/par-test.cc]: Yes but's that incorrect and the compiler may throw warnings.

**Comment by costashatz** [comment]
> I think it should be tested first. Could you at least add libtbb-dev to the debian packages installed in .travis.yml ?

(1) If we do not enable TBB, nothing changes anywhere.
(2) We already agreed on making a different PR for travis reformatting (and appveyor support) and testing TBB there. We can fix any possible issues there.
(3) In our lab, we are making heavy usage of libcmaes with TBB enabled every day. It works on *Ubuntu and Mac OS.

File: tests/par-test.cc (hunk 13 above)
Comment (costashatz) [reply tests/par-test.cc]: > Yes but's that incorrect and the compiler may throw warnings.

Fixed

**Comment by jschueller** [comment]
I dont want a broken master and "fix any possible issues" later.
It's better if you prove that it works and enable tbb please, that should not be so hard to add "--enable-tbb" to the configure line ?

**Comment by jschueller** [comment]
Why removing it ? If you add an optional dependency it's better if tested.

**Comment by costashatz** [comment]
> Why removing it ? If you add an optional dependency it's better if tested.

We need to create variants in travis to test with or without TBB, with or without OpenMP (this is not tested for the moment for example), etc. Because the code might work with TBB enabled, but something would have broken when TBB is not enabled or not found. This is out of the scope of this PR (it needs quite some refactoring in the travis file and some testing). So, I will create a new PR that handles all these cases.

In the meantime, to convince you that this PR works, I can create a commit to enable tbb, let it run on travis and after it succeeds (and you are convinced that works), I will remove it and make the new PR (that I just described above) to handle all the travis variants.

OK?

**Comment by costashatz** [comment]
You can see an example of a rather complicated but complete travis file for different variants at this [link](https://github.com/resibots/limbo/blob/master/.travis.yml). I am planning to do something similar with libcmaes to test with different configurations (with/without TBB, with/without OpenMP etc)..

**Comment by jschueller** [comment]
I the mean time you could just test with tbb enabled and disabled: 
```
- ./configure --enable-python LDFLAGS="-L/usr/local/lib/"
- make
- ./configure --enable-python  --with-tbb LDFLAGS="-L/usr/local/lib/"
- make
```

**Comment by costashatz** [comment]
The results of the travis tests:

- The compilation with and without TBB is fine (see the 2 linux runs) and thus this PR does not break anything.
- The 2 runs for OSX errored because they timed-out.

So, I suggest to revert to the original travis file (i.e., no TBB) to prevent the OSX runs from timing-out (and the master branch from appearing broken when it is not). Shortly after this PR is merged, I will make the new PR with the proper travis variants so that no run times-out. What do you think?

**Comment by beniz** [comment]
@costashatz thanks for the good work, and @jschueller thanks for the thorough review. You guys are really making this lib better!

It seems it is close to a merge. My remarks at this stage:

- I've struggled a bit to get `tbb` detected on Ubuntu 17.04: lib is in `/usr/lib/x86_64-linux-gnu/` and is named `libtbb.so.2`, so by default the `configure` script does not detect it. Below is my little custom hack in order to get it detected, unfortunately this is not generic enough:
```
diff --git a/configure.ac b/configure.ac
index 8b36ed1..e5a6f90 100644
--- a/configure.ac
+++ b/configure.ac
@@ -62,10 +62,10 @@ AC_HELP_STRING([--with-tbb=DIR], [Path to the TBB]),
 AC_MSG_CHECKING([for tbb])
 
 lib_tbb_avail=no
-AC_CHECK_FILE($tbb_dirs/include/tbb/tbb.h,[lib_tbb_avail=yes],
+AC_CHECK_FILE(/usr/include/tbb/tbb.h,[lib_tbb_avail=yes],
        [AC_CHECK_FILE(/usr/local/include/tbb/tbb.h,[tbb_dirs=/usr/local])])
-AC_CHECK_FILE($tbb_dirs/lib/libtbb.so,[lib_tbb_avail=yes],
-       [AC_CHECK_FILE(/usr/local/lib/libtbb.so,[tbb_dirs=/usr/local],[lib_tbb_avail=no])])
+AC_CHECK_FILE($tbb_dirs/libtbb.so.2,[lib_tbb_avail=yes],
+       [AC_CHECK_FILE(/usr/local/lib/libtbb.so.2,[tbb_dirs=/usr/],[lib_tbb_avail=no])])
 
 if test x"$lib_tbb_avail" = xyes; then
    AC_SUBST([TBB_INC], [$tbb_dirs/include])
```

- @costashatz would have some tangible / demonstrative results on parallelization performance with TBB vs OpenMP ? Giving users a reason to use one or the other is a good practice. Runtime on some of the test functions, on a a multi-core machine would be great.

I've restarted an OSX build on Travis in order to see whether the timeout is still here.

**Comment by costashatz** [comment]
> lib is in /usr/lib/x86_64-linux-gnu/ and is named libtbb.so.2

This is very weird. There should always exist a file `libtbb.so`; I, too, have `libtbb.so` and `libtbb.so.2`. Make sure that tbb is installed correctly. I will also try in newer Ubuntu versions soon.

> so by default the configure script does not detect it

Regarding this issue, I would vote for removing the Makefiles completely from the library (not in this PR, but a new one). CMake generates makefiles/visual studio projects (and a lot more) and is much easier to use (although my personal preference in automated build-systems is not CMake, it is one of the most widely used). It is especially easier to write package/libraries detection scripts than Makefiles.

Of course, I can (and I will) test some quick n' dirty hacks for making TBB detectable both in previous Ubuntu versions and newer ones, but in the long term I strongly vote for removing completely the Makefiles.

> would have some tangible / demonstrative results on parallelization performance with TBB vs OpenMP ? Giving users a reason to use one or the other is a good practice. Runtime on some of the test functions, on a a multi-core machine would be great.

Honestly, I do not believe that there will be a significant difference in performance between TBB and OpenMP. There are however some points of TBB that makes it appealing (to me):

- Works on any compiler and any OS; just a library to include/link to
- Automatic/Dynamic scheduling (no need to worry about anything)

Of course, I can create some plots in some test functions on a multi-core computer. I will come back at you when I have something..
